server/src/main/java/org/elasticsearch/index/IndexVersion.java

```diff
@@ -245,6 +251,10 @@ static Collection<IndexVersion> getAllVersions() {
         return VERSION_IDS.values();
     }
 
+    public static IndexVersion readVersion(StreamInput in) throws IOException {
+        return fromId(in.readVInt());
```

Comment:
nit: do we need wire compatibility with Version for IndexVersion? Maybe not since it is not used in the handshake? If not, I suggest we use read/write Int right away, as a varint here is more complicated to read when we know the value is always 4 bytes.